### PR TITLE
Add v1 plugin register function into v2 plugin template

### LINKED ISSUE #44022: Let the plugin create template add stubs to both versions of the plugins API

### PR DESCRIPTION
**Pull request by xster**
## Description

Let the ecosystem plugin developer continue to support both embeddings for now. 

## Related Issues

Fixes https://github.com/flutter/flutter/issues/44022

## Tests

Devicelab tests added in plugin_test

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by xster** [comment]
Can feature flags / flutter config be out-of-process injectable per flutter process invocation? Such as via https://github.com/flutter/flutter/blob/master/dev/devicelab/lib/tasks/plugin_tests.dart

**Comment by xster** [comment]
Oh wait wooooot, it already is https://github.com/flutter/flutter/blob/9093cf15f631b354c0fae4d726a8638b308a5e31/packages/flutter_tools/lib/src/features.dart#L142. Man you guys already thought of everything :D

File: dev/devicelab/lib/tasks/plugin_tests.dart [hunk 1]
```diff
@@ -26,10 +26,12 @@ TaskFunction combine(List<TaskFunction> tasks) {
 /// Defines task that creates new Flutter project, adds a local and remote
 /// plugin, and then builds the specified [buildTarget].
 class PluginTest {
-  PluginTest(this.buildTarget, this.options);
+  PluginTest(this.buildTarget, this.options, { this.pluginCreateEnvs, this.appCreateEnvs });
 
   final String buildTarget;
   final List<String> options;
+  final Map<String, String> pluginCreateEnvs;
```
Comment (blasten) [inline dev/devicelab/lib/tasks/plugin_tests.dart]: nit: `Envs` -> `Environment`

Comment (blasten) [inline dev/devicelab/lib/tasks/plugin_tests.dart]: nit: `Envs` -> `Environment`

Comment (xster) [reply dev/devicelab/lib/tasks/plugin_tests.dart]: sg

File: packages/flutter_tools/templates/plugin/android-java.tmpl/src/main/java/androidIdentifier/pluginClass.java.tmpl [hunk 2]
```diff
@@ -21,6 +22,20 @@ public class {{pluginClass}} implements FlutterPlugin, MethodCallHandler {
...
 
+  // This function is optional and equivalent to onAttachedToEngine. It supports the old
+  // pre-Flutter-1.12 Android projects. Plugin developers are encouraged to continue supporting
+  // plugin registration via this function while apps migrate to use the new Android APIs
+  // post-flutter-1.12.
+  //
+  // Developers are encouraged to share logic between onAttachedToEngine and registerWith to keep
+  // them functionaly equivalent. Only one of onAttachedToEngine or registerWith will be called
+  // depending on the user's project. onAttachedToEngine or registerWith must both be defined
+  // in the same class.
```
Comment (blasten) [inline packages/flutter_tools/templates/plugin/android-java.tmpl/src/main/java/androidIdentifier/pluginClass.java.tmpl]: @sfshaza2 do you prefer to use third-person or second-person when referring to developers in docs?

**Comment by xster** [comment]
Also cc @InMatrix for usability.

File: packages/flutter_tools/templates/plugin/android-java.tmpl/src/main/java/androidIdentifier/pluginClass.java.tmpl [hunk 3]
```diff
@@ -21,6 +22,20 @@ public class {{pluginClass}} implements FlutterPlugin, MethodCallHandler {
     channel.setMethodCallHandler(new {{pluginClass}}());
   }
 
+  // This function is optional and equivalent to onAttachedToEngine. It supports the old
+  // pre-Flutter-1.12 Android projects. Plugin developers are encouraged to continue supporting
+  // plugin registration via this function while apps migrate to use the new Android APIs
+  // post-flutter-1.12.
+  //
+  // Developers are encouraged to share logic between onAttachedToEngine and registerWith to keep
```
Comment (matthew-carroll) [inline packages/flutter_tools/templates/plugin/android-java.tmpl/src/main/java/androidIdentifier/pluginClass.java.tmpl]: Should we link to a canonical example or a guide?

**Review by matthew-carroll — APPROVED** [review]
LGTM

File: packages/flutter_tools/templates/plugin/android-java.tmpl/src/main/java/androidIdentifier/pluginClass.java.tmpl (hunk 3 above)
Comment (mklim) [reply packages/flutter_tools/templates/plugin/android-java.tmpl/src/main/java/androidIdentifier/pluginClass.java.tmpl]: +1, but we don't have a guide here that's good enough to link to yet. The Wiki is for Flutter developers ourselves and not Flutter users. The guide on our wiki is great for our purposes but we'd need to move it to the website and probably clean it up some first. I think writing a comprehensive tutorial would take some time.

File: packages/flutter_tools/templates/plugin/android-java.tmpl/src/main/java/androidIdentifier/pluginClass.java.tmpl (hunk 2 above)
Comment (mklim) [reply packages/flutter_tools/templates/plugin/android-java.tmpl/src/main/java/androidIdentifier/pluginClass.java.tmpl]: Generally I think we prefer second person. https://developers.google.com/style/person

**Review by mklim — APPROVED** [review]
LGTM.

This is aside from this PR's main point so I'm not blocking the patch on it, but should we be defaulting to creating a v2 embedding at all right now in the template? If I were a user who created a template I'd go to look for how to use the v2 embedding, and then not find anything obvious, and then probably be stuck. It's not going to be easy for us to write a straightforward guide on how to migrate to v2 because so many of the APIs that a plugin author may need are in (seemingly) random revisions of Flutter and not on stable. Even among the Flutter team there has been a lot of ongoing changes in how to do this.

File: packages/flutter_tools/templates/plugin/android-java.tmpl/src/main/java/androidIdentifier/pluginClass.java.tmpl (hunk 2 above)
Comment (xster) [reply packages/flutter_tools/templates/plugin/android-java.tmpl/src/main/java/androidIdentifier/pluginClass.java.tmpl]: https://github.com/flutter/website/pull/3007#discussion_r325219526 oh you're right. Shams actually pointed this out to me before. Thanks

File: packages/flutter_tools/templates/plugin/android-java.tmpl/src/main/java/androidIdentifier/pluginClass.java.tmpl (hunk 3 above)
Comment (xster) [reply packages/flutter_tools/templates/plugin/android-java.tmpl/src/main/java/androidIdentifier/pluginClass.java.tmpl]: re: examples

If this is about how to share logic, I think it's generic enough that it doesn't need to be explained in a code comment. And the template code already does it. 

If it's about how to migrate an old plugin to a new plugin, we should document it in https://github.com/flutter/flutter/issues/44224 and https://github.com/flutter/flutter/issues/44225. Though I don't think we need to link those in this file. The users just made a new plugin here. They're not trying to migrate a plugin at this moment. 

@mklim, comment may or may not be orthogonal. https://github.com/flutter/flutter/issues/44224 we should create a proper flutter.dev guide for the migration. Though we don't need to do it this week.

**Comment by xster** [comment]
@mklim, for migrations https://github.com/flutter/flutter/issues/44224, we should document it. But as in the comment above, this template is for new plugin developers. And nothing in https://flutter.dev/docs/development/packages-and-plugins/developing-packages points to APIs in the old plugin registry specifically. So they'll have as much or as little instructions they always had developing plugins. https://github.com/flutter/flutter/issues/44225 is also for mapping equivalents between APIs.

File: packages/flutter_tools/templates/plugin/android-java.tmpl/src/main/java/androidIdentifier/pluginClass.java.tmpl [hunk 4]
```diff
@@ -21,6 +22,20 @@ public class {{pluginClass}} implements FlutterPlugin, MethodCallHandler {
     channel.setMethodCallHandler(new {{pluginClass}}());
   }
 
+  // This function is optional and equivalent to onAttachedToEngine. It supports the old
+  // pre-Flutter-1.12 Android projects. Plugin developers are encouraged to continue supporting
+  // plugin registration via this function while apps migrate to use the new Android APIs
+  // post-flutter-1.12.
```
[the remark this answers is not in the document]
Comment (xster) [reply packages/flutter_tools/templates/plugin/android-java.tmpl/src/main/java/androidIdentifier/pluginClass.java.tmpl]: I think versions are the source of truth and dates are a proxy. We should be as specific as possible.

**Review by blasten — APPROVED** [review]

**Comment by mklim** [comment]
> And nothing in https://flutter.dev/docs/development/packages-and-plugins/developing-packages points to APIs in the old plugin registry specifically. 

@xster that's fair. I think my concern is coming from the added complexity here.

In addition to knowing how to do all the v1 work, the plugin author needs to register the different lifecycle callbacks and implement a few different interfaces to get all the objects they may want. There will be some detective work involved in figuring out what those are and why they want those objects. And those lifecycle callbacks and the static `registerWith` method need to call through to the same logic in the end so that it all stays compatible. And the callbacks need to unregister their listeners appropriately so there aren't any leaks (the template code as written is leaking the MethodCallHandler). And some of the methods on master don't exist on stable. Etc. The Javadocs on all the v2 stuff is helpful but a higher level guide just to tell devs where to look and what to be careful about relying on would be good too.

I don't think this being a new template vs a migration really matters that much for the complexity, if a dev is still supporting both. I could be overestimating it though.

**Comment by xster** [comment]
Discussed offline. We'll document that complexity in https://github.com/flutter/flutter/issues/44236 and https://github.com/flutter/flutter/issues/44225

**Pull request merged**